File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/714-best-time-to-buy-and-sell-stock-with-transaction-fee.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function(prices, fee) {
    let noStock = 0, stock = -Infinity
    
    for (let price of prices) {
        let tempNoStock = Math.max(noStock, stock + price)
        let tempStock = Math.max(stock, (noStock - price - fee))
        
        noStock = tempNoStock
        stock = tempStock
    }
    
    return noStock
};

File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/714-best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int noStock = 0, stock = Integer.MIN_VALUE;
        int tempNoStock, tempStock;

        for (int price : prices) {
            tempNoStock = Math.max(noStock, stock + price);
            tempStock = Math.max(stock, (noStock - price - fee));

            noStock = tempNoStock;
            stock = tempStock;
        }

        return noStock;
    }
}